Added speed+duplex to Interface model (updated UI)

nrnvgh opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes #6979 
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
-->

Added speed/duplex attributes to `Interface` object with tests for same. This version has smarter UI vs the previous version and adds the following improvements:

- Custom user-supplied speeds (specified in Kbps)
- Advanced filtering now supports filtering by speed gt/gte/lt/lte a target value

## Notes on the implementation

- The new `speed_widget.html` follows the basic pattern established by `Circuit.commit_rate`, with a tiny bit of javascript.
- Contents of `interface_edit.html` were moved to the new `interface_create.html` and the former updated to inherit from the latter, following the pattern seen elsewhere.
- The `{% block title %}` in `interface_create` has no special provision for 'creating' because there's no title text of create today. Compare:
    - [Creating interface UI @ demo.nautobot](https://demo.nautobot.com/dcim/interfaces/add/?device=43e0c7a2-3939-5fcf-bc2f-c659f1c40d46&return_url=/dcim/devices/43e0c7a2-3939-5fcf-bc2f-c659f1c40d46/interfaces/)
    - [Editing interface UI @ demo.nautobot](https://demo.nautobot.com/dcim/interfaces/89c73c2b-3758-4743-af1c-739e43e5c86a/edit/?return_url=/dcim/devices/43e0c7a2-3939-5fcf-bc2f-c659f1c40d46/interfaces/?tab=interfaces)

# Screenshots
<!--
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->

https://github.com/user-attachments/assets/096d38b3-6701-4f1d-a349-346d5068a720


# TODO
<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

nrnvgh committed:
Closes #6979: Add Interface speed/duplex

    - Model: add fields and type-aware validation; speed stored as Integer (Kbps)
    - Forms: include fields in create/edit/bulk; TypedChoiceField coercion; widgets; filters for speed/duplex.
    - API: serializer fields (IntegerField Kbps, ChoiceField duplex); CSV import/export round-trip covered.
    - GraphQL: expose speed as Int (Kbps) and duplex as String
    - UI: list/detail render humanized speed
    - Docs: update interface.md for fields, validation rules, units
    - Tests: add model, form, filter, API, and GraphQL tests validating constraints and behavior.

nrnvgh committed:
Closes #6979: Add Interface speed/duplex

    - Model: add fields and type-aware validation; speed stored as Integer (Kbps)
    - Forms: include fields in create/edit/bulk; TypedChoiceField coercion; widgets; filters for speed/duplex.
    - API: serializer fields (IntegerField Kbps, ChoiceField duplex); CSV import/export round-trip covered.
    - GraphQL: expose speed as Int (Kbps) and duplex as String
    - UI: list/detail render humanized speed
    - Docs: update interface.md for fields, validation rules, units
    - Tests: add model, form, filter, API, and GraphQL tests validating constraints and behavior.

nrnvgh committed:
markdown cleanup

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Re-trigger pipeline

nrnvgh committed:
Re-trigger pipeline #2

nrnvgh committed:
added testcase to confirm validation allows LAG interface with no speed or duplex

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Updated humanize_speed to use decimalized strings

Examples:

    - 1544 => "1.544 Mbps"
    - 100000 => "100 Mbps"
    - 10000000 => "10 Gbps"
    - 1000000000 => "1 Tbs"
    - 1600000000 => "1.6 Tbps"
    - 10000000000 => "10 Tbps"

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speed

nrnvgh committed:
Added .changed fragment

nrnvgh committed:
Update nautobot/core/templatetags/helpers.py

Fixed docstring

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

nrnvgh committed:
Updated choices name for 1.6Tbps to be consistent with other entries and humanize_speed philosophy

nrnvgh committed:
Make interface speed/duplex choices to nautobot.apps.choices

nrnvgh committed:
refactored _validate_speed_and_duplex to reduce the number of return points

nrnvgh committed:
removed commented-out 'speed' attribute

nrnvgh committed:
removed unneeded speed/duplex from graphql/types.py

nrnvgh committed:
ruff cleanup

nrnvgh committed:
fixed case-based test failure

nrnvgh committed:
removed debugging code

nrnvgh committed:
synced with main dev branch

nrnvgh committed:
removed debug comments

nrnvgh committed:
removed old code

nrnvgh committed:
new template files

nrnvgh committed:
Added some table render test cases

nrnvgh committed:
added label to 'speed' attr

nrnvgh committed:
ruff cleanup

nrnvgh committed:
whitespace cleanup

nrnvgh committed:
removed 'in' from loop_expr comparison since it would never happen

nrnvgh committed:
moved new multvalue+number filter to nautobot.core.filters

nrnvgh committed:
made the new filter available to apps

nrnvgh committed:
djhtml linting

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speedduplex-advancedui

glennmatthews commented on nautobot/core/utils/filtering.py:
```diff
@@ -121,7 +121,15 @@ def get_filterset_parameter_form_field(model, parameter, filterset=None):
+            form_field = forms.MultipleChoiceField(
+                choices=field.choices,
+                widget=StaticSelect2Multiple,
+            )
```
Not sure about this - what if I want to filter on `exact=12345` for whatever reason and 12345 isn't one of the choices?

glennmatthews commented on nautobot/dcim/templates/dcim/interface_create.html:
```diff
@@ -0,0 +1,79 @@
+                                <li class="text-danger">{{ error }}</li>
+                            {% endfor %}
+                        </ul>
+                    {% endif %}
```
Same comment here about creating a custom widget and using that.

nrnvgh replied:
agreed; updated.

nrnvgh committed:
Update changes/6979.changed per suggestion

Co-authored-by: Joe Wesch <10467633+joewesch@users.noreply.github.com>

nrnvgh committed:
simplified conditionals in _validate_speed_and_duplex

nrnvgh committed:
updated InterfaceBulkUpdate to use a widget instead of full templates

nrnvgh committed:
updated InterfaceCreateView to use new widget instead of full templates

nrnvgh commented on nautobot/dcim/templates/dcim/interface_bulk_edit.html:
```diff
@@ -0,0 +1,94 @@
+                                        <div class="col-md-9">
+                                            <div class="input-group">
+                                                {{ field }}
+                                                {% include 'dcim/inc/speed_widget.html' with target_field='speed' btn_class='btn-sm' %}
```
updated.

nrnvgh committed:
djhtml fixup

glennmatthews reviewed (commented):
Nearly there I think!

nrnvgh committed:
removed now-unneeded speed_widget files

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speedduplex-advancedui

joewesch reviewed (commented):
LGTM once you've addressed Glenn's feedback

nrnvgh committed:
updated filter fields to accept dropdown or freeform values

nrnvgh committed:
Added speed/duplex to InterfaceTemplate

- Constants:
  - Add COPPER_TWISTED_PAIR_IFACE_TYPES (BASE‑T types).

- Models:
  - InterfaceTemplate: add speed (Kbps) and duplex; validate:
    - speed/duplex not allowed for LAG/virtual/wireless.
    - duplex only for copper twisted‑pair types.
    - propagate speed/duplex in create_instance().
  - Interface: use COPPER_TWISTED_PAIR_IFACE_TYPES for duplex validation; clarify docstring.

- Forms:
  - InterfaceTemplateForm/Create/BulkEdit: include speed (IntegerField, Kbps) and duplex; use NumberWithSelect for speed; StaticSelect2 for duplex; labels/help as needed.
  - Import form includes speed/duplex.

- Tables:
  - InterfaceTemplateTable: add speed/duplex columns; humanize speed; update field ordering/defaults.

- Migrations:
  - Add speed/duplex to interface template; remove superseded migration; add consolidated migration.

- Tests:
  - Update/add API/models/tables/views tests for speed/duplex creation, rendering, and validation (including duplex rejection on non‑twisted‑pair types).

nrnvgh committed:
Fixed behavior of 'speed' in bulkinterfaceedit

    When 'set null' was checked, the field would disappear but the drop-down arrow would remain. This
    update causes the field not to fade out, which was far and away the easiest path and is consistent
    with other UI elements in the bulkinterfaceedit page.

nrnvgh committed:
ruff fixes to migration file

nrnvgh committed:
Made MultiValueNumberFilter choices aware and got rid of MultiValueNumberWithChoicesFilter (which was created for this patch)

nrnvgh committed:
Merge branch 'develop' into u/nrnvgh-6979-interface-speedduplex-advancedui

nrnvgh committed:
re-factored number_input do use data-* field names

nrnvgh committed:
Removed 'help_text' from 'speed' model attributes

    - With it in place, create forms were rendering with a field name, a placeholder text in the field, /and/
      a caption under the field, all with same value.

nrnvgh commented:
Last of the commits is in, unless there's additional feedback from these changes. One thing to note: as implemented, when a user selects a value from a dropdown in an interface create/edit field, the integer value is placed in the field. In filters, the human-readable label is placed there (unless it's not a drop-down option). I'm leaning strongly towards changing the latter, just ran out of time.

glennmatthews commented on nautobot/dcim/tests/test_models.py:
```diff
@@ -726,6 +726,101 @@ def test_interface_template_sets_interface_status(self):
+                        type=if_type,
+                        speed=InterfaceSpeedChoices.SPEED_1G,
+                    ).full_clean()
+                    self.assertIn("Speed is not applicable to this interface type.", str(cm.exception))
```
```suggestion
                self.assertIn("Speed is not applicable to this interface type.", str(cm.exception))
```

glennmatthews reviewed (commented):
Couple of quick comments on the latest changes, will re-review in total next.

nrnvgh committed:
Update nautobot/dcim/tests/test_models.py

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>
